tls-tutorial/doc/tls_introduction.md

```diff
@@ -0,0 +1,222 @@
+
+#### TLS Handshake
+
+Before Alice and Bob can start their actual conversation, they first need to establish the secure communication. This is called the ‘TLS handshake’.
+
+Alice and Bob add random data and a timestamp to the first messages of the communication, in order to prevent so called ‘replay attacks’. In a replay attack, Eve records a communication and tries to take over the next one by replaying the recorded messages. Since the random data and timestamp change each time, Eve’s attack fails.
+
+Now the communication partners establish each other’s authenticity &mdash; or more precisely they should. We’ll get to that in a minute. For now, you just need to be aware that the goal of this step is to make sure that the communication partners are who they claim to be. The driving technology behind this step is called public-key cryptography or asymmetric cryptography.
+
+Think of a  box with a lock that needs two keys &mdash; one for locking it and one for unlocking it. Bob can give copies of the locking key to anyone he likes, Alice for example. As long as Bob keeps the other key, only he can open the box to get out what other people put in.
+
+In cryptographic terms we speak of a private key and a public key. As the name suggests, Bob needs to keep the private key secret and can share the other one. Bob could theoretically post the public key on Twitter without compromising security. But there are better ways to do this, and we will discuss them later.
+
+Asymmetric cryptography works nicely for authenticating a communication partner. Conceptually, Alice creates a piece of data that is sufficiently hard to guess (for example a large random number) and asks Bob to encrypt it with that Bob’s private key. Bob then sends the result back. Alice uses Bob’s public key to decrypt the data and compare it with the original. If the data matches, Alice knows she is talking to Bob. The details are, of course, more complicated in reality. But as a first approximation this example is good enough.
+
+### Key Exchange
+
+One aspect that makes all mechanisms like TLS more complex is the fact that asymmetric cryptography is expensive. This means it takes up a large amount of computing resources, especially CPU and network bandwidth. On the other hand, we have no resources to waste, especially when it comes to long-distance communication. So the TLS inventors came up with the concept of switching to symmetric cryptography.
+
+As said, asymmetric cryptography is too expensive to transfer mass data. What’s a more efficient way then? Negotiating a new key after the initial handshake and switching the encryption to symmetric (e.g. [AES](http://www.faqs.org/rfcs/rfc3268.html)) from that point on. So, symmetric encryption is by orders of magnitude, cheaper.
+
+The symmetric key is what we call the 'session key' as it is created each time you start a new session. In fact, it can even be exchanged during a running session if it reaches a predefined expiry.
```

Comment:
```suggestion
The symmetric key is what we call the 'session key' as it is created each time you start a new session. In fact, it can even be replaced during a running session if it reaches a predefined expiry.
```